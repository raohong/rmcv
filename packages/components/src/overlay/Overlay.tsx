import { easings } from '@react-spring/web';
import { useLockScroll, useMergeRefs } from '@rmc-vant/hooks';
import { styled } from '@rmc-vant/system';
import classNames from 'classnames';
import React from 'react';
import { baseStyleReset } from '../_styles';
import Animation from '../animation';
import { useThemeProps } from '../config-provider';
import Portal from '../portal';
import { OverlayName, overlayClassNames } from './classNames';
import type { OverlayProps } from './interface';

const OverlayRoot = styled(Animation, {
  name: OverlayName,
  slot: 'root',
  overridesResolver: () => ['root'],
})(baseStyleReset, ({ theme }) => ({
  position: 'fixed',
  left: 0,
  top: 0,
  bottom: 0,
  right: 0,
  touchAction: 'none',
  zIndex: theme.zIndex.overlay,
  backgroundColor: 'rgba(0, 0, 0, 0.7)',
}));

const Overlay = React.forwardRef<HTMLDivElement, OverlayProps>((props, ref) => {
  const {
    open,
    className,
    style,
    onClick,
    children,
    lazyRender,
    teleport,
    zIndex,
    duration,
    lockScroll = true,
    ...rest
  } = useThemeProps(OverlayName, props);
  const lockRef = useLockScroll(!!open, !lockScroll);
  const domRef = useMergeRefs(ref, lockRef);

  const enabled = duration === undefined || duration > 0;

  return (
    <Portal teleport={teleport} disablePortal={!teleport}>
      <OverlayRoot
        componentState={{ open }}
        onClick={onClick}
        className={classNames(overlayClassNames.root, className)}
        aria-hidden={!open}
        immediate={!enabled}
        ref={domRef}
        config={
          duration
            ? {
                duration,
                easing: easings.easeInCubic,
              }
            : undefined
        }
        from={{
          opacity: 0,
        }}
        enter={{
          opacity: 1,
        }}
        leave={{
          opacity: 0,
        }}
        style={{ zIndex, ...style }}
        animate={open}
        forceRender={!lazyRender}
        {...rest}
      >
        {children}
      </OverlayRoot>
    </Portal>
  );
});

export default Overlay;
