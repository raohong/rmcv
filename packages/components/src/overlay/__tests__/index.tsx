import { render, screen } from '@testing-library/react';
import React from 'react';
import Overlay from '../Overlay';

test('render correctly', () => {
  const tree = render(<Overlay />);

  expect(tree.asFragment()).toMatchSnapshot();
});

test('render with zIndex', () => {
  render(<Overlay data-testid="test" zIndex={2} />);

  expect(screen.getByTestId('test')).toHaveStyle({
    'z-index': 2,
  });
});

test('render with visible', async () => {
  render(<Overlay visible data-testid="test" />);

  expect(screen.getByTestId('test')).toHaveAttribute('aria-hidden', 'false');
});

test('render with duration', async () => {
  const duration = 6;

  render(<Overlay visible duration={duration} data-testid="test" />);

  expect(screen.getByTestId('test')).toHaveStyle({
    'animation-duration': '6s',
  });
});
