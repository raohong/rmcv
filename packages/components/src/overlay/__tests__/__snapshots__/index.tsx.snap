// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render correctly 1`] = `
<DocumentFragment>
  <div
    class="rmc-vant-overlay"
    style="z-index: 1; opacity: 0; display: none;"
  />
</DocumentFragment>
`;
